encoders/csv.go

```diff
@@ -1,11 +1,50 @@
 package encoders
 
-import "github.com/jszwec/csvutil"
+import (
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/jszwec/csvutil"
+	"io"
+)
+
+var _ runtime.Marshaler = (*csvEncoder)(nil)
 
 // csvEncoder implements Marshaller for the CSV format.
 type csvEncoder struct {
 }
 
+// NewDecoder returns a Decoder which reads byte sequence from "r".
+func (e csvEncoder) NewDecoder(r io.Reader) runtime.Decoder {
+	return runtime.DecoderFunc(func(v interface{}) error {
+		raw, err := io.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		return e.Unmarshal(raw, v)
+	})
```

Comment:
This same exact implementation is shared by the `raw` implementation, and I'm sure it will be shared by others as the steps are fairly generic. Let's extract this and reuse it across implementations.

Also, we should try casting `r` to `Closer` and defer a `Close()` call if it casts to prevent memory leaks.

```suggestion
	return runtime.DecoderFunc(func(v interface{}) error {
		raw, err := io.ReadAll(r)
		if err != nil {
			return err
		}
		if c, ok := r.(io.Closer); ok {
			defer c.Close()
		}
		return e.Unmarshal(raw, v)
	})
```

Reply:
Closing the body is a responsibility from an upper layer, which is usually handled by the HTTP server. I don't think we should move forward with this change.

Reply:
I meant the suggestion, not extracting the repeated logic.

Reply:
Closing the body is typically the responsibility of whoever reads the body, which in this case is this function given that we're calling `io.ReadAll`. I would agree with your point if this function received the contents of the reader instead of the reader itself.

Reply:
Makes sense. Added the `Close` call. Thanks!